src/java.desktop/share/classes/java/awt/geom/Path2D.java

```diff
@@ -2129,6 +2081,140 @@ public final Rectangle getBounds() {
+            double endX, endY;
+            switch (type) {
+                case PathIterator.SEG_MOVETO, PathIterator.SEG_LINETO:
+                    endX = coords[0];
+                    endY = coords[1];
+                    break;
+                case PathIterator.SEG_QUADTO:
+                    endX = coords[2];
+                    endY = coords[3];
+                    break;
+                case PathIterator.SEG_CUBICTO:
+                    endX = coords[4];
+                    endY = coords[5];
+                    break;
+                default:
+                    continue pathIteratorLoop;
+            }
+
+            if (isEmpty) {
+                // we're seeding our bounds for the first time:
+                isEmpty = false;
+                leftX = rightX = endX;
+                topY = bottomY = endY;
+            } else {
+                // extend our rectangle to cover the point at t = 1:
+                leftX = (endX < leftX) ? endX : leftX;
+                rightX = (endX > rightX) ? endX : rightX;
+                topY = (endY < topY) ? endY : topY;
+                bottomY = (endY > bottomY) ? endY : bottomY;
+            }
+
+            // here's the slightly trickier part: examine quadratic and cubic
+            // segments for extrema where t is between (0, 1):
+
+            boolean definedParametricEquations;
+            if (type == PathIterator.SEG_QUADTO) {
+                definedParametricEquations = true;
+
+                x_coeff[3] = 0.0;
+                x_coeff[2] = lastX - 2.0 * coords[0] + coords[2];
+                x_coeff[1] = -2.0 * lastX + 2.0 * coords[0];
+                x_coeff[0] = lastX;
+
+                y_coeff[3] = 0;
+                y_coeff[2] = lastY - 2.0 * coords[1] + coords[3];
+                y_coeff[1] = -2.0 * lastY + 2.0 * coords[1];
+                y_coeff[0] = lastY;
+            } else if (type == PathIterator.SEG_CUBICTO) {
+                definedParametricEquations = true;
+
+                x_coeff[3] = -lastX + 3.0 * coords[0] - 3.0 * coords[2] + coords[4];
+                x_coeff[2] = 3.0 * lastX - 6.0 * coords[0] + 3.0 * coords[2];
+                x_coeff[1] = -3.0 * lastX + 3.0 * coords[0];
+                x_coeff[0] = lastX;
+
+                y_coeff[3] = -lastY + 3.0 * coords[1] - 3.0 * coords[3] + coords[5];
+                y_coeff[2] = 3.0 * lastY - 6.0 * coords[1] + 3.0 * coords[3];
+                y_coeff[1] = -3.0 * lastY + 3.0 * coords[1];
+                y_coeff[0] = lastY;
+            } else {
+                definedParametricEquations = false;
+            }
+
+            if (definedParametricEquations) {
+                int tExtremaCount = Curve.findExtrema(x_coeff, tExtrema);
+                for(int i = 0; i < tExtremaCount; i++) {
+                    double t = tExtrema[i];
+                    if (t > 0 && t < 1) {
+                        double x = x_coeff[0] + t * (x_coeff[1] + t * (x_coeff[2] + t * x_coeff[3]));
```

Comment:
using 3rd order polynom is only useful for cubic curves, for quads 2nd order is enough.
How to improve precision on (abcd) or (bcd) polynomial evaluation ?

Reply:
Ideally the compensated-horner scheme should be used to guarantee optimal precision (2x slower):
paper: 
https://www-pequan.lip6.fr/~jmc/polycopies/Compensation-horner.pdf

See julia code: 
https://discourse.julialang.org/t/more-accurate-evalpoly/45932/6

Reply:
I don't think I can implement that on my own. Even if I fully understood the julia code, I assume I'd want to use the Math.fma(..) method. That method uses BigDecimals, which I assume (?) we don't want to use in a loop like this for performance reasons.

If this level of high-precision is considered essential to resolving this bug: then personally I'm at an impasse and we should either close this PR or I'll need some external help.

Or possible compromises might include:

A. Add a method like this in Curve.java:
```
public double evaluateCubicPolynomial(double a, double b, double c, double d, double t) {
    return d + t * (c + t * (b + t * a)));
}
```

And add a ticket that someone else can address separately to improve the accuracy of that method. (Also maybe other areas in the AWT codebase should use the same method? Like `Order3#XforT`?)

B. We can modify my new method signature from:
`public static Rectangle2D getBounds2D(PathIterator pi)`
to:
`public static Rectangle2D getBounds2D(PathIterator pi, boolean highPrecision)`

Where if `highPrecision` is true we use the new implementation *which may suffer rounding errors and be a few ulps too small*. And if it is false then we use the old implementation which suffers from the original bug. This has the benefit of definitely not breaking any existing code in existence. The downside is it puts the responsibility on developers to learn about and implement this new method. But at least it would exist.

Thoughts?

Reply:
I agree it is out of the scope of this issue to implement high-precision polynom evaluation (compensated horder scheme), that I or anybody else could implement it later.
To ensure bounds2D are enclosing the shape entirely, I propose to add a small margin (10 ulps or more) to ascertain precision is good enough.
To determine the precision level, using fuzzy testing is the good approach : generate lots ofrandom paths => check bounds2D accuracy in ulp units.

Reply:
@prrace @mrserb what do you think on this bug in terms of precision requirements?
- should bbox always include control points ?
- should ideal shape always fit in bbox ? So I propose to add a margin > numerical inaccuracies.

I can develop a numerical approach based on fuzzy testing:
- generate random curves (high magnitude changes on control points)
- use both this algorithm (double) and implement the same with BigDecimal
- estimate the extrema error = max distance(pts big decimal, pts double)

Reply:
Ideally jdk could provide new math methods:
- evalPoly(eqn[])
- find roots...
Based on compensated horner scheme and Math.fma() ~ 0.5 ulp

Reply:
I attempted to explore a numerical approach to pad the bounding box here 4b9d87d6d03e923bb2663fc8f4f77b7549df6e70 ... but I don't like the results so far. Any suggestions? (Or if you have a better approach in mind we can discard that entire class...)

Reply:
You made an amazing job !
Your bigdecimal impl looks good.
I will play with your test experiment... asap.
I think numerical accuracies are related to the dynamic /magnitude of values: points.
The test should evaluate(max error) depending on the quick length(curve) ~ manhattan norm(curve arms), so small curves have less numerical error whereas huge curve (10^50 length) means points are wide spread and coeffs... solver...points are less accurate.

I will try making sampled control points vary in log scale: 10^-6 to 10^38 (float max) so curve length will vary in huge range and determine the histogram of max(error) / length ratio.

Reply:
see condition number = magnitude range on such plot:
https://github.com/JuliaMath/AccurateArithmetic.jl/blob/master/test/figs/sum_accuracy.svg

Reply:
I just pushed a commit ( 40bda06 ) that addresses the machine error problem to my satisfaction. It includes a unit test that failed before that commit and now passes.

At your convenience let me know your thoughts. (That commit continues to focus on very small doubles; it may (?) need help to address very large double errors.)

Reply:
That looks interesting as the new margin looks like ulp(x - bbox), however it may be not enough as the position error depends on curve polynom coefficients (magnitude).

I derived your (yesterday) test to play with the margin and try small to huge curves. 
I found the condition number ~ sum (abs(coeffs)) so margin = math.ulp(cond). But sometimes the quad solver is giving t with few ulps error that cause high position shifts.

I will go on testing your last approach and share my experiments

Reply:
As this accuracy problem is a more general issue on quad / cubic curves, I created the Marlin-Math project to ascertain the math function accuracy.
See https://github.com/bourgesl/marlin-math/blob/main/src/main/java/test/FindExtremaAccuracyTest.java

I will propose a new upper margin value that quarantees the edge condition.